libkbfs/unflushed_path_cache.go

```diff
@@ -0,0 +1,328 @@
+func (upc *unflushedPathCache) startInitializeOrWait(ctx context.Context) (
+	bool, error) {
+	// Retry in case the original initializer has to abort due to
+	// error; limit the number of retries by the lifetime of `ctx`.
+	for {
+		doInit, readyCh := upc.doStartInitialization()
+		if doInit {
+			return true, nil
+		} else if readyCh == nil {
+			// Already initialized.
+			return false, nil
+		}
+		select {
+		case <-readyCh:
+			continue
+		case <-ctx.Done():
+			return false, ctx.Err()
+		}
+	}
+}
+
+func (upc *unflushedPathCache) abortInitialization() {
+	upc.lock.Lock()
+	defer upc.lock.Unlock()
+	upc.state = upcUninitialized
+	upc.appendQueue = nil
+	upc.removeQueue = nil
+	close(upc.ready)
+	upc.ready = nil
+}
+
+// addUnflushedPaths populates the given unflushed paths object.  The
+// caller should NOT be holding any locks, as it's possible that
+// blocks will need to be fetched.
+func addUnflushedPaths(ctx context.Context,
+	uid keybase1.UID, kid keybase1.KID, codec kbfscodec.Codec,
+	log logger.Logger, irmds []ImmutableRootMetadata, cpp chainsPathPopulator,
+	unflushedPaths unflushedPathsMap) error {
+	// Make chains over the entire range to get the unflushed files.
+	chains := newCRChainsEmpty()
+	processedOne := false
+	for _, irmd := range irmds {
+		if _, ok := unflushedPaths[irmd.Revision()]; ok {
+			if processedOne {
+				return fmt.Errorf("Couldn't skip revision %d after "+
+					"already processing one", irmd.Revision())
+			}
+
+			log.CDebugf(ctx, "Skipping unflushed paths for revision %d "+
+				"since it's already in the cache", irmd.Revision())
+			continue
+		}
+
+		processedOne = true
+		winfo := writerInfo{
+			uid:      uid,
+			kid:      kid,
+			revision: irmd.Revision(),
+			// There won't be any conflicts, so no need for the
+			// username/devicename.
+		}
+		err := chains.addOps(codec, irmd.data, winfo, irmd.localTimestamp)
+		if err != nil {
+			return err
+		}
+	}
+	if !processedOne {
+		return nil
+	}
+
+	chains.mostRecentMD = irmds[len(irmds)-1]
+
+	err := cpp.populateChainPaths(ctx, log, chains, true)
+	if err != nil {
+		return err
+	}
+
+	for _, chain := range chains.byOriginal {
+		for _, op := range chain.ops {
+			revPaths := unflushedPaths[op.getWriterInfo().revision]
+			if revPaths == nil {
+				revPaths = make(map[string]bool)
+				unflushedPaths[op.getWriterInfo().revision] = revPaths
+			}
+			revPaths[op.getFinalPath().String()] = true
+		}
+	}
+	return nil
+}
+
+// prepUnflushedPaths returns a set of paths that were updated in the
+// given revision.
+func (upc *unflushedPathCache) prepUnflushedPaths(ctx context.Context,
+	uid keybase1.UID, kid keybase1.KID, codec kbfscodec.Codec,
+	log logger.Logger, irmd ImmutableRootMetadata) (
+	unflushedPathsPerRevMap, error) {
+	cpp := func() chainsPathPopulator {
+		upc.lock.Lock()
+		defer upc.lock.Unlock()
+		return upc.chainsPopulator
+	}()
+
+	// The unflushed paths haven't been initialized yet.
+	if cpp == nil {
```

Comment:
On further thought, we probably can't get rid of the different states, because even if we try to populate the cache when creating the `tlfJournal`, it might be too large to do so quickly, and we'll have to return incomplete results until it is flushed enough to be populated quickly.  But I think it would still be a win (in a future PR) to be able to get `cpp` from the `KBFSOps` interface and at least try to initialize it at construction time.